Piwik.Tracker.Tests/PiwikTrackerTests.cs

```diff
@@ -1,62 +1,591 @@
+            var actual = _sut.GetCustomVariable(variableId, variableScope);
+            //Assert
+            Assert.That(actual, Is.Null);
+        }
 
         [Test]
-        [TestCase(Scopes.Page, null, null, null)]
         [TestCase(Scopes.Page, 2, null, null)]
         [TestCase(Scopes.Page, 2, "myPageVar", "myPageVarValue")]
-        [TestCase(Scopes.Event, null, null, null)]
         [TestCase(Scopes.Event, 3, null, null)]
         [TestCase(Scopes.Event, 3, "myEventVar", "myEventVarValue")]
-        [TestCase(Scopes.Visit, null, null, null)]
         [TestCase(Scopes.Visit, 4, null, null)]
         [TestCase(Scopes.Visit, 4, "myVisitVar", "myVisitVarValue")]
-        public void GetCustomVariable_Test(Scopes variableScope, int? variableId, string variableName, string variableValue)
+        public void GetCustomVariable_WhenVariableIsSet_ReturnsCorrectVariable(Scopes variableScope, int variableId, string variableName, string variableValue)
+        {
+            //Arrange
+            _sut.SetCustomVariable(variableId, variableName, variableValue, variableScope);
+
+            //Act
+            var actual = _sut.GetCustomVariable(variableId, variableScope);
+            //Assert
+            Assert.That(actual.Name, Is.EqualTo(variableName));
+            Assert.That(actual.Value, Is.EqualTo(variableValue));
+        }
+
+        [Test]
+        public void GetCustomVariable_WhenInvalidScopeArgument_Throws()
         {
             //Arrange
-            var sut = new PiwikTracker(SiteId, PiwikBaseUrl);
-            if (variableId != null)
+            //Act & Assert
+            Assert.Throws<ArgumentException>(() => _sut.GetCustomVariable(1, (Scopes)1234));
+        }
+
+        [Test]
+        public void Ctor_WhenNoUrlProvided_Throws()
+        {
+            Assert.Throws<ArgumentException>(() => new PiwikTracker(0, null));
+        }
+
+        [Test]
+        [TestCase("testCharset", "&cs=testCharset")]
+        [TestCase("utf16", "&cs=utf16")]
+        [TestCase("", null)]
+        public void SetPageCharset_WhenSpecified_IsAddedToRequest(string charset, string expected)
+        {
+            // Arrange, Act
+            _sut.SetPageCharset(charset);
+            // Assert
+            var actual = _sut.GetRequest(SiteId);
+            if (string.IsNullOrEmpty(expected))
             {
-                sut.SetCustomVariable(variableId.Value, variableName, variableValue, variableScope);
+                Assert.That(actual, Does.Not.Contain("&cs="));
             }
-            //Act
+            else
+            {
+                Assert.That(actual, Does.Contain(expected));
+            }
+        }
 
-            var actual = sut.GetCustomVariable(variableId ?? 99, variableScope);
-            //Assert
-            if (variableId != null)
+        [Test]
+        [TestCase("http://myurl.com/index.html")]
+        [TestCase("")]
+        public void SetUrl_WhenSpecified_IsAddedToRequest(string pageUrl)
```

Comment:
I appreciate your point, but having a unified approach to testing requests :
 - helps having a clear view of what is currently being tested and which tests are missing
 - makes the code easier to understand
 - facilitates the work of the developer by removing one question : "should I use approach A or approach B?" sometimes there is no clear way to make a decision and this is when it gets messy

In this specific case, it would also remove the need to mark a substantial amount of attributes internal which does seem to look like a [code smell](https://softwareengineering.stackexchange.com/questions/161619/unit-testing-its-a-code-smell-if-youre-refactoring-and-there-are-no-collabor).

The fact that it takes a couple seconds more to run is a price I think is fair to pay for all these benefits.

Also to be clear, I am not implying that all tests should use the mocked http server.

An example of tests that could live independently is the one that unit test along several axis formatting methods such as FormatDateValue, FormatTimestamp. In this case I would suggest introducing a new class to group all these methods and test it independently of the rest.